src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs

```diff
@@ -105,40 +109,40 @@ public TextAndVersion GetValue(LoadTextOptions options, CancellationToken cancel
+            return (RecoverableText)_initialSourceOrRecoverableText;
         }
 
-        public async Task<TextAndVersion> GetValueAsync(LoadTextOptions options, CancellationToken cancellationToken)
+        public TextAndVersion GetValue(LoadTextOptions options, CancellationToken cancellationToken)
         {
-            if (_initialSourceOrRecoverableText is ITextAndVersionSource source)
-            {
-                // replace initial source with recoverable text if it hasn't been replaced already:
-                Interlocked.CompareExchange(
-                    ref _initialSourceOrRecoverableText,
-                    value: new RecoverableText(source, await source.GetValueAsync(options, cancellationToken).ConfigureAwait(false), options, _services),
-                    comparand: source);
-            }
+#pragma warning disable CA2012 // Use ValueTasks correctly
+            var valueTask = GetRecoverableTextAsync(useAsync: false, options, cancellationToken);
+#pragma warning restore CA2012 // Use ValueTasks correctly
+            Contract.ThrowIfFalse(valueTask.IsCompleted, "GetRecoverableTextAsync should have completed synchronously since we passed 'useAsync: false'");
+            var recoverableText = valueTask.GetAwaiter().GetResult();
 
-            // If we have a recoverable text but the options it was created for do not match the current options
-            // and the initial source supports reloading, reload and replace the recoverable text.
-            var recoverableText = (RecoverableText)_initialSourceOrRecoverableText;
-            if (recoverableText.LoadTextOptions != options && recoverableText.InitialSource != null)
-            {
-                Interlocked.Exchange(
-                    ref _initialSourceOrRecoverableText,
-                    new RecoverableText(recoverableText.InitialSource, await recoverableText.InitialSource.GetValueAsync(options, cancellationToken).ConfigureAwait(false), options, _services));
-            }
+            return recoverableText.ToTextAndVersion(recoverableText.GetValue(cancellationToken));
+        }
 
-            recoverableText = (RecoverableText)_initialSourceOrRecoverableText;
+        public async Task<TextAndVersion> GetValueAsync(LoadTextOptions options, CancellationToken cancellationToken)
+        {
+            var recoverableText = await GetRecoverableTextAsync(useAsync: true, options, cancellationToken).ConfigureAwait(false);
             return recoverableText.ToTextAndVersion(await recoverableText.GetValueAsync(cancellationToken).ConfigureAwait(false));
         }
 
+        public async Task<VersionStamp> GetVersionAsync(LoadTextOptions options, CancellationToken cancellationToken)
+        {
+            var recoverableText = await GetRecoverableTextAsync(useAsync: true, options, cancellationToken).ConfigureAwait(false);
+            return recoverableText.Version;
```

Comment:
this is the crux of the change.  the current code calls through `GetValueAsync` which then forces `recoverableText.GetValueAsync` (which goes back to reload if dumped from memory).  This one doesn't need to go back to do anything.  The RecoverableText instance already stores the VersionStamp directly on it.